base/genericmemory.jl

```diff
@@ -0,0 +1,223 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+## genericmemory.jl: Managed Memory
+
+"""
+    GenericMemory{isatomic, T} <: AbstractVector{T}
```

Comment:
Just a minor note that the name `GenericMemory` doesn't really make it obvious that this subtypes `AbstractArray`. Perhaps that's intentional (i.e. we don't want people to treat these too strongly as "arrays").

Reply:
I really think this shouldn't subtype `AbstractVector`. I doubt you want to use a "named" `GenericMemory` as input to all kinds of linear algebra functions etc. This should IMO be a very "basic" type with not many methods defined on it.

Reply:
> the question is just, what to do with existing uses of AbstractVector?

Nothing? All I am saying is that I don't think the existing generic methods written for `AbstractArray` will be very useful for `GenericMemory`. Even though you can implement `size` and `getindex`, it doesn't mean you *have* to make it an `AbstractArray` because you then opt into things that you might not want. 

Many methods on `AbstractArray` return an `Array` unless you specialize them but it seems that is rarely something you want to happen if you have a `GenericMemory`. So people will start overloading these methods to return a `GenericMemory` etc and this just doesn't seem like something you want to deal with.

Reply:
That makes sense, yes. I'm on board with keeping this as its own thing, and having actual array semantics on different types on top of this.

Reply:
No, this is all very intentional that it is an AbstractVector, not an accident or pun. In fact, it is much more descriptive of a mathematical vector than Vector was, since Vector adds lots of features as "general collection of objects" which are not mathematically-oriented and do not exist for Memory. However, Vector will remain the preferred type for methods to return in most cases, because Vector is more general. That is also intentional and not something that should be changed in most of those algorithms.

Reply:
Fair enough - I guess we'll see how much code expects to `push!` on `AbstractVector` or similar, should it leak out.

Reply:
Most other AbstractArray types do not support `push!`, so I don't see that being an issue

Reply:
> In fact, it is much more descriptive of a mathematical vector than Vector was, since Vector adds lots of features as "general collection of objects" which are not mathematically-oriented and do not exist for Memory. 

You want to have a basic memory type that is free from even being thought about in the same way as a mathematical vector. If you want you can wrap something on top of `Memory` that you consider an `AbstractVector` but the core memory type should not be passable into `ForwardDiff.gradient` and to `.+` and a bunch of other stuff that doesn't make sense for it.

Reply:
the counterpoint is that given that it has a size, and getindex, you will get sensible results out of all of those.

Reply:
Only to answer the last question, which may answer most of the others, they should be taking a DenseArray already, so no changes are needed for them. (Except that reminds me that this should really be a DenseVector, and not merely an AbstractVector. Usually I don't bother to express that level of granularity in array APIs, so it had slipped my mind.)

Reply:
> which may answer most of the others

I don't really see how the other questions are answered from this.